File: iOS/TUILiveKit/Source/State/Operation/BattleState.swift
//
//  BattleState.swift
//  TUILiveKit
//
//  Created by krabyu on 2024/8/26.
//

import Foundation
import RTCRoomEngine

let battleDuration: TimeInterval = 30
let battleRequestTimeout: TimeInterval = 10
let battleEndInfoDuration: TimeInterval = 5

struct BattleState: Encodable {
    
    var battleId: String = ""
    var battleUsers: [BattleUser] = []
    var sentBattleRequests: [BattleUser] = []
    var receivedBattleRequest: BattleUser?
    var isBattleRunning: Bool = false
    var battleConfig: BattleConfig = BattleConfig()
    var needResponse: Bool = true
    
    var isInWaiting: Bool = false
    var isOnDisplayResult: Bool = false
}

struct BattleUser: Codable {
    var roomId: String
    var userId: String
    var avatarUrl: String
    var userName: String
    var score: UInt
    var ranking: Int
    
    init() {
        roomId = ""
        userId = ""
        avatarUrl = ""
        userName = ""
        score = 0
        ranking = 1
    }
    
    init(_ battleUser: TUIBattleUser) {
        self.roomId = battleUser.roomId
        self.userId = battleUser.userId
        self.userName = battleUser.userName
        self.avatarUrl = battleUser.avatarUrl
        self.score = battleUser.score
        ranking = 1
    }
}

extension BattleUser: Equatable {
    static func == (lhs: BattleUser, rhs: BattleUser) -> Bool {
        return lhs.userId == rhs.userId && lhs.roomId == rhs.roomId && lhs.score == rhs.score
    }
}

extension BattleUser: Hashable {
    
    func hash(into hasher: inout Hasher) {
        hasher.combine(roomId)
    }
}

struct BattleConfig: Encodable {
    var duration: TimeInterval
    var needResponse: Bool
    var extensionInfo: String
    
    init() {
        duration = 30
        needResponse = true
        extensionInfo = ""
    }
    
    init(_ config: TUIBattleConfig) {
        self.duration = config.duration
        self.needResponse = config.needResponse
        self.extensionInfo = config.extensionInfo
    }
}

File: Flutter/livekit/lib/live_stream/state/battle_state.dart
import 'package:flutter/material.dart';
import 'package:rtc_room_engine/api/extension/tui_live_battle_manager.dart';

class LSBattleState {
  static const int battleDuration = 30;
  static const int battleRequestTime = 10;
  static const int battleEndInfoDuration = 5;

  final ValueNotifier<String> battleId = ValueNotifier('');
  final ValueNotifier<List<BattleUser>> battleUsers = ValueNotifier([]);
  final ValueNotifier<ReceivedBattleRequest?> receivedBattleRequest = ValueNotifier(null);
  final ValueNotifier<int> durationCountDown = ValueNotifier(0);
  BattleConfig battleConfig = BattleConfig();
  bool needResponse = true;

  final ValueNotifier<bool> isInWaiting = ValueNotifier(false);
  bool isShowingStartWidget = false;
  final ValueNotifier<bool> isBattleRunning = ValueNotifier(false);
  final ValueNotifier<bool> isOnDisplayResult = ValueNotifier(false);
}

typedef ReceivedBattleRequest = (String battleId, BattleUser inviter);

class BattleUser {
  String roomId;
  String userId;
  String avatarUrl;
  String userName;
  int score;
  int ranking;

  BattleUser(
      {this.roomId = '',
      this.userId = '',
      this.avatarUrl = '',
      this.userName = '',
      this.score = 0,
      this.ranking = 1});

  BattleUser.fromTUIBattleUser(TUIBattleUser battleUser)
      : roomId = battleUser.roomId,
        userId = battleUser.userId,
        avatarUrl = battleUser.avatarUrl,
        userName = battleUser.userName,
        score = battleUser.score,
        ranking = 1;

  @override
  String toString() {
    return 'BattleUser{roomId: $roomId, userId: $userId, avatarUrl: $avatarUrl, userName: $userName, score: $score, ranking: $ranking}';
  }
}

class BattleConfig {
  int duration;
  bool needResponse;
  String extensionInfo;

  BattleConfig(
      {this.duration = 30, this.needResponse = true, this.extensionInfo = ''});
}
